scipy/stats/_continuous_distns.py

```diff
@@ -3791,6 +3796,72 @@ def fit(self, data, *args, **kwds):
             fshape_s = fshape_n / fscale
         return fshape_s, floc, fscale
 
+    def _ppf_isf(self, q, mu, upper):
+        """ calculate ppf or isf depending on the value of `upper`.
+        If upper is True, then compute isf, else ppf. This implementation is
+        based on Gyner & Smyth (2016) and uses newton's method with carefully
+        selected starting points that guarantee convergence regardless of the
+        probability."""
+        x0 = np.empty(np.broadcast(q, mu).shape)
+        q = np.broadcast_to(q, x0.shape)
+        mu = np.broadcast_to(mu, x0.shape)
+
+        # Select the start values for Newton's Method using a quantile of
+        # the gamma (if right tail is required) or normal distribution
+        # when `q` is very small.
+        mask = q < 1e-05
+        if upper:
+            mu3 = mu[mask] ** 3
+            x0[mask] = sc.gammainccinv(1 / mu3, q[mask]) / mu3
+        else:
+            x0[mask] = mu[mask] / (_norm_ppf(q[mask]) ** 2)
+
+        # when `q` is not very small the initial value for Newton's method is
+        # the mode of the distribution. If `k` is large, then we use equation 3
+        # of Giner & Smyth (2016) to prevent subtractive cancellation.
+        # Note that the threshold of 100 picked here is arbitrary, a much
+        # bigger value could be chosen if necessary.
+        k = 1.5 * mu[~mask]
+        x0[~mask] = _lazywhere(
+            mu[~mask] > 100,
+            (mu[~mask], k),
+            lambda mu, k: mu * (0.5 / k - 0.125 / (k ** 3) + 0.0625 / (k**6)),
+            f2=lambda mu, k: mu * (np.sqrt(1 + k * k) - k)
+        )
+
+        logF = self._logsf if upper else self._logcdf
+        sign = -1 if upper else 1
+        delta = np.empty_like(x0)
+        lq = np.log(q)
+
+        for _ in range(100):
+            lx = logF(x0, mu)
+            lp = self._logpdf(x0, mu)
+            # to avoid subtractive cancellation when `d` is very small we use a
+            # taylor series expansion to approximate (p - p(xn)) as shown in
+            # page 8 of Giner & Smyth (2016)
+            d = lq - lx
+            mask = np.abs(d) < 1e-05
+            delta[mask] = (
+                d[mask] * np.exp(lq[mask] + np.log1p(-0.5*d[mask]) - lp[mask])
+            )
+            delta[~mask] = (
+                q[~mask] * np.exp(-lp[~mask]) - np.exp(lx[~mask] - lp[~mask])
+            )
+
+            x = x0 + sign * delta
+            if np.allclose(x, x0, atol=1e-08):
```

Comment:
Does he paper say what they used as the convergence criterion? This looks a bit coarse.

Reply:
No the paper does not mention it. It is really implementation dependent. Do you have a different suggestion? Your second point is addressed in [`31e909a` (#13665)](https://github.com/scipy/scipy/pull/13665/commits/31e909ad3e764000fb2c1a8333e24d4cf934687f)